File: deploy/audit_backend/processors/audit_api_client_test.py
# python3
# Copyright 2019 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Tests for the (place-holder) AuditApiClient."""

import datetime
import textwrap
from absl.testing import absltest

from deploy.audit_backend.processors import audit_api_client
from deploy.audit_backend.processors import audit_api_resources


def get_fake_now():
  """Returns a fake now function that increases by 1 day on each call."""
  next_time = datetime.datetime(2019, 1, 1, 8, 30, 55)

  def now():
    nonlocal next_time
    current_time = next_time
    next_time = current_time + datetime.timedelta(days=1)
    return current_time

  return now


class AuditApiClientTest(absltest.TestCase):

  def setUp(self):
    super(AuditApiClientTest, self).setUp()
    self.api_client = audit_api_client.AuditApiClient(now=get_fake_now())

  def test_get_latest_config_md5_hash(self):
    for i in range(5):
      self.api_client.create_new_project_config({'config_num': i},
                                                'hash_{}'.format(i))
    self.assertEqual('hash_4', self.api_client.get_latest_config_md5_hash())

  def test_get_latest_config_md5_hash_no_configs(self):
    self.assertIsNone(self.api_client.get_latest_config_md5_hash())

  def test_create_new_project_config(self):
    for i in range(5):
      self.api_client.create_new_project_config({'config_num': i},
                                                'hash_{}'.format(i))
    for i in range(5):
      config, scanners = self.api_client._saved_configs[i]
      expected_time = datetime.datetime(2019, 1, 1 + i, 8, 30, 55)
      expected_config = audit_api_resources.ProjectConfig(
          name='projectConfigs/{}'.format(int(expected_time.timestamp())),
          create_time=expected_time.isoformat(),
          config={'config_num': i},
          hash='hash_{}'.format(i))
      self.assertEqual(expected_config, config)
      self.assertEqual({}, scanners)

  def test_debug_clear(self):
    for i in range(5):
      self.api_client.create_new_project_config({'config_num': i},
                                                'hash_{}'.format(i))
    self.api_client.debug_clear()
    self.assertEqual([], self.api_client._saved_configs)

  def test_debug_dump_all(self):
    for i in range(2):
      self.api_client.create_new_project_config({'config_num': i},
                                                'hash_{}'.format(i))
    expected = textwrap.dedent("""\
    configs:
    - name: projectConfigs/1546360255
      create_time: 2019-01-01T08:30:55
      hash: hash_0
    - name: projectConfigs/1546446655
      create_time: 2019-01-02T08:30:55
      hash: hash_1""")
    self.assertEqual(expected, self.api_client.debug_dump_all())


if __name__ == '__main__':
  absltest.main()

File: deploy/collector/processors/audit_api_client_test.go
/*
 * Copyright 2019 Google LLC.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package processors

import (
	"strings"
	"testing"
	"time"

	"google3/third_party/golang/godebug/pretty/pretty"
)

var (
	configText1 = "config#1"
	configText2 = "config#2"
	createTime1 = time.Now()
	createTime2 = createTime1.Add(time.Hour)
	md5Hash1    = "hash1"
	md5Hash2    = "hash2"

	auditConfig1 = &AuditConfig{
		projectConfigText: configText1,
		createTime:        &createTime1,
		md5Hash:           md5Hash1,
	}

	auditConfig2 = &AuditConfig{
		projectConfigText: configText2,
		createTime:        &createTime2,
		md5Hash:           md5Hash2,
	}
)

func TestAuditConfigs(t *testing.T) {
	ac := NewAuditAPIClient()

	ac.CreateAuditConfig(auditConfig1)
	ac.CreateAuditConfig(auditConfig2)

	config, err := ac.FetchLatestAuditConfig()
	if err != nil {
		t.Errorf("FetchLatestAuditConfig return error: %v", err)
	}
	if diff := pretty.Compare(config.projectConfigText, configText2); diff != "" {
		t.Errorf("FetchLatestAuditConfig() returned if (-want +got): \n%s", diff)
	}
	if diff := pretty.Compare(ac.DebugDumpAuditConfigs(), strings.Join([]string{configText1, configText2}, "|")); diff != "" {
		t.Errorf("DebugDumpAuditConfigs() returned if (-want +got): \n%s", diff)
	}

	ac.DebugClearAuditConfigs()
	if diff := pretty.Compare(ac.DebugDumpAuditConfigs(), ""); diff != "" {
		t.Errorf("DebugDumpAuditConfigs() returned if (-want +got): \n%s", diff)
	}
}
